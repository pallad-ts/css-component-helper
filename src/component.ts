import {variantNameToDOMPropertyName} from "./variantNameToDOMPropertyName";
import {dashCase} from "./dashCase";
import {VariantsDescriptor} from "./VariantsDescriptor";
import {StyleObject} from "./StyleObject";
import {Variant} from "./Variant";
import {ComponentReference} from "./ComponentReference";
import {CSSVariable} from "./CSSVariable";
import type * as CSS from 'csstype';
import {CSSComponent} from "./CSSComponent";

function flyweightInitializer<T>(initializer: (name: string) => T) {
	const map = new Map<string, T>();
	return (key: string): T => {
		if (map.has(key)) {
			return map.get(key)!;
		}

		const value = initializer(key);
		map.set(key, value);
		return value;
	}
};

function ourStylesToCssProperties({vars, ...styles}: StyleObject.CSSTypeProperties): CSS.Properties {
	if (vars) {
		for (const [varName, value] of Object.entries(vars)) {
			// @ts-ignore
			styles[varName] = value;
		}
	}
	return styles as CSS.Properties;
}

export function component<TVariants extends VariantsDescriptor, TVariables extends string = string, TSubComponentName extends string = string>(name: string) {

	const variantInitializer = flyweightInitializer(name => {
		return new Variant(name, variantNameToDOMPropertyName(dashCaseName, name));
	});

	const componentReferenceInitializer = flyweightInitializer(name => {
		return new ComponentReference(`${dashCaseName}-${dashCase(name)}`);
	});

	const variableInitializer = flyweightInitializer(name => {
		return new CSSVariable(`--${dashCaseName}-${dashCase(name)}`);
	});

	const dashCaseName = dashCase(name);
	const o: CSSComponent<TVariants, TVariables, TSubComponentName> = {
		name,
		dashCaseName,
		styleObject(object) {
			const root: any = {
				...(object.base ? ourStylesToCssProperties(object.base) : undefined),
			};

			if (object.variants) {
				for (const [variant, definition] of Object.entries(object.variants)) {
					if (!definition) {
						continue;
					}

					const {base, ...variantValueDefinitions} = definition;
					if (base) {
						root[o.variant(variant).selector.self] = ourStylesToCssProperties(base);
					}

					for (const [variantValue, definition] of Object.entries(variantValueDefinitions)) {
						root[o.variant(variant).selector.withValue(variantValue as any).self] = ourStylesToCssProperties(definition as StyleObject.CSSTypeProperties);
					}
				}
			}
			return root;
		},
		variable(name) {
			return variableInitializer(name);
		},
		variant<T extends VariantsDescriptor.Names<TVariants>>(name: T) {
			return variantInitializer(name) as Variant<TVariants[T]>;
		},
		applyVariants(value: TVariants) {
			return Object.fromEntries(
				Object.entries(value)
					.map(([variant, value]) => {
						return o.variant(variant).entry(value as any);
					})
			);
		},
		subComponentRef(name) {
			return componentReferenceInitializer(name);
		}
	};

	return o;
}
