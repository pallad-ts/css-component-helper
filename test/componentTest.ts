import {component as cssComponent} from "@src/component";

describe('component', () => {
	const component = cssComponent<{
		size?: 'small' | 'medium' | 'large',
		isLoading?: boolean,
		index: number;
	}, 'padding' | 'isDisabled', 'end' | 'start'>('button');

	describe('style object', () => {
		it('general', () => {
			expect(component.styleObject({
				base: {
					vars: {
						isDisabled: '1'
					},
					border: '1px solid red',
				},
				variants: {
					size: {
						base: {
							content: '"size"',
						},
						small: {
							vars: {
								padding: '10',
							},
							content: '"size-small"',
						},
						medium: {
							content: '"size-medium"',
						},
						large: {
							content: '"size-large"'
						}
					}
				}
			}))
				.toMatchSnapshot();
		});
	})

	describe('applying variants', () => {
		it('general', () => {
			expect(component.applyVariants({
				index: 10,
				size: 'small',
				isLoading: true
			}))
				.toMatchSnapshot()
		});
	});

	describe('component reference', () => {
		it('getting ref returns always the same object', () => {
			const ref1 = component.subComponentRef('end');
			const ref2 = component.subComponentRef('end');

			expect(ref1)
				.toBe(ref2);
		});

		it('general', () => {
			expect(component.subComponentRef('end'))
				.toMatchSnapshot();
		});
	});

	describe('variable', () => {
		it('general', () => {
			expect(component.variable('padding'))
				.toMatchSnapshot();
		});
	});
});
