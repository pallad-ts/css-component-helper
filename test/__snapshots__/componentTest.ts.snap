// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component applying variants general 1`] = `
{
  "data-button-index": 10,
  "data-button-is-loading": true,
  "data-button-size": "small",
}
`;

exports[`component component reference general 1`] = `
ComponentReference {
  "entry": [
    "data-ref",
    "button-end",
  ],
  "name": "button-end",
  "object": {
    "data-ref": "button-end",
  },
  "selector": CSSBaseSelector {
    "raw": "[data-ref="button-end"",
    "self": "&[data-ref="button-end"",
  },
}
`;

exports[`component style object general 1`] = `
{
  "&[data-button-size="large"]": {
    "content": ""size-large"",
  },
  "&[data-button-size="medium"]": {
    "content": ""size-medium"",
  },
  "&[data-button-size="small"]": {
    "--button-padding": "10",
    "content": ""size-small"",
  },
  "&[data-button-size]": {
    "content": ""size"",
  },
  "--button-is-disabled": "1",
  "border": "1px solid red",
}
`;

exports[`component variable general 1`] = `
CSSVariable {
  "name": "--button-padding",
  "reference": "var(--button-padding)",
}
`;
